### PR TITLE
Allow explicit NWB file paths in CLI and API

### LINKED ISSUE #21: Adjust CLI
ATM CLI is

```bash
nwb2bids <directory of NWB files> <BIDS organized directory>
```

which is quite "restrictive" as it requires all nwb files to be in a folder or on the other side -- all NWB files would be considered.  I think it would be better to get it

```bash
nwb2bids -o <BIDS organized directory> <paths>
```

where `<paths>` is 1 or more of either 
- `NWB file`
- `directory of NWB files`

that would also allow for 
- easy chaining of invocation with "paths producing" helper via `xargs` , e.g. `find -iname *fordandi*.nwb | xargs nwb2bids -o ds000xxx`

### PR DESCRIPTION
**Pull request by CodyCBakerPhD**
fix #21

### CODE REVIEW

**Comment by CodyCBakerPhD** [comment]
Should probably come after #22

File: src/nwb2bids/_converters/_dataset_converter.py [hunk 1]
```diff
@@ -53,6 +66,20 @@ def from_nwb_directory(
         dataset_converter = cls(session_converters=session_converters, dataset_description=dataset_description)
         return dataset_converter
 
+    # TODO: remove
```
Comment (yarikoptic) [inline src/nwb2bids/_converters/_dataset_converter.py:69]: remove what?

File: src/nwb2bids/_converters/_dataset_converter.py [hunk 2]
```diff
@@ -53,6 +66,20 @@ def from_nwb_directory(
...
+    @classmethod
+    @pydantic.validate_call
+    def from_nwb_directory(
+        cls, nwb_directory: pydantic.DirectoryPath, additional_metadata_file_path: pydantic.FilePath | None = None
+    ) -> typing_extensions.Self:
+        message = "The method `.from_nwb_directory` is deprecated. Please use `.from_nwb` instead."
```
Comment (yarikoptic) [inline src/nwb2bids/_converters/_dataset_converter.py:75]: IMHO at this age of development I would not bother about migrations/deprecations , especially at python API level

File: src/nwb2bids/_converters/_dataset_converter.py [hunk 3]
```diff
@@ -119,7 +146,9 @@ def write_participants_metadata(self, bids_directory: str | pathlib.Path) -> Non
         bids_directory : directory path
             The path to the directory where the BIDS dataset will be created.
         """
-        self._establish_bids_directory_and_check_metadata(bids_directory=bids_directory)
+        bids_directory = bids_directory = self._establish_bids_directory_and_check_metadata(
```
Comment (yarikoptic) [inline src/nwb2bids/_converters/_dataset_converter.py:149]:
```suggestion
        bids_directory = self._establish_bids_directory_and_check_metadata(
```

File: src/nwb2bids/_converters/_session_converter.py [hunk 4]
```diff
@@ -195,11 +216,14 @@ def write_events_files(self, bids_directory: str | pathlib.Path) -> None:
 
         # IDEA: check events.json files, see if all are the same and if so remove the duplicates and move to outer level
 
-    def _establish_bids_directory_and_check_metadata(self, bids_directory: pathlib.Path) -> None:
+    def _establish_bids_directory_and_check_metadata(self, bids_directory: pathlib.Path) -> pathlib.Path:
```
Comment (yarikoptic) [inline src/nwb2bids/_converters/_session_converter.py:219]:
```suggestion
    def _establish_bids_directory_and_check_metadata(self, bids_directory: str|pathlib.Path) -> pathlib.Path:
```

**Review by yarikoptic — CHANGES_REQUESTED** [review]
minor comments

File: src/nwb2bids/_converters/_dataset_converter.py (hunk 1 above)
Comment (CodyCBakerPhD) [reply src/nwb2bids/_converters/_dataset_converter.py]: 'nothing' anymore 😉

File: src/nwb2bids/_converters/_session_converter.py (hunk 4 above)
Comment (CodyCBakerPhD) [reply src/nwb2bids/_converters/_session_converter.py]: Handled in previous PR

File: src/nwb2bids/_command_line_interface/_main.py [hunk 5]
```diff
@@ -90,18 +91,26 @@ def _run_convert_nwb_dataset(
...
     Convert NWB files to BIDS format.
 
-    NWB_DIRECTORY : The path to the folder containing NWB files.
+    NWB : Either a path to a folder containing NWB files, or an explicit list of NWB file paths.
     """
+    nwb_directory = None
+    nwb_file_paths = None
+    if pathlib.Path(nwb[0]).is_dir():
+        nwb_directory = pathlib.Path(nwb[0])
+    else:
```
[the remark this answers is not in the document]
Comment (CodyCBakerPhD) [reply src/nwb2bids/_command_line_interface/_main.py:112]: Added a better message and test case

File: src/nwb2bids/_command_line_interface/_main.py [hunk 6]
```diff
@@ -90,18 +91,26 @@ def _run_convert_nwb_dataset(
...
+        nwb_file_paths = nwb
```
[the remark this answers is not in the document]
Comment (CodyCBakerPhD) [reply src/nwb2bids/_command_line_interface/_main.py]: IDK I mean, Pydantic will check this when passed to the API. Would rather not bother with it myself

**Comment by CodyCBakerPhD** [comment]
@candleindark Just a reminder to take a look when you get a chance

File: src/nwb2bids/_converters/_dataset_converter.py [hunk 7]
```diff
@@ -24,26 +24,38 @@ class DatasetConverter(BaseConverter):
 
     @classmethod
     @pydantic.validate_call
-    def from_nwb_directory(
-        cls, nwb_directory: pydantic.DirectoryPath, additional_metadata_file_path: pydantic.FilePath | None = None
+    def from_nwb(
```
[the remark this answers is not in the document]
Comment (CodyCBakerPhD) [reply src/nwb2bids/_converters/_dataset_converter.py]: Done

**Comment by CodyCBakerPhD** [comment]
@candleindark All changes discussed from yesterday have been made

File: tests/integration/test_convert_nwb_dataset.py [hunk 8]
```diff
@@ -80,36 +80,84 @@ def test_ecephys_convert_nwb_dataset(ecephys_nwbfile_path: pathlib.Path, tempora
...
+                "files": {
+                    "sub-123_ses-456_ecephys.nwb",
+                },
+            },
+        }
+        nwb2bids.testing.assert_subdirectory_structure(
+            directory=new_bids_directory, expected_structure=expected_structure
+        )
+    finally:
+        os.chdir(path=initial_working_directory)
```
Comment (candleindark) [inline tests/integration/test_convert_nwb_dataset.py:163]:
```suggestion
        os.chdir(path=temporary_bids_directory)
    try:
        new_bids_directory = temporary_bids_directory / "bids"

        nwb_paths = [minimal_nwbfile_path]
        nwb2bids.convert_nwb_dataset(nwb_paths=nwb_paths)

        expected_structure = {
            new_bids_directory: {
                "directories": {"sub-123"},
                "files": {"dataset_description.json", "participants.json", "participants.tsv"},
            },
            new_bids_directory
            / "sub-123": {
                "directories": {"ses-456"},
                "files": {"sub-123_sessions.json", "sub-123_sessions.tsv"},
            },
            new_bids_directory
            / "sub-123"
            / "ses-456": {
                "directories": {"ecephys"},
                "files": set(),
            },
            new_bids_directory
            / "sub-123"
            / "ses-456"
            / "ecephys": {
                "directories": set(),
                "files": {
                    "sub-123_ses-456_ecephys.nwb",
                },
            },
        }
        nwb2bids.testing.assert_subdirectory_structure(
            directory=new_bids_directory, expected_structure=expected_structure
        )
    finally:
        os.chdir(path=initial_working_directory)
```

The above or just use [`monkeypatch`](https://docs.pytest.org/en/stable/reference/reference.html#pytest.MonkeyPatch.chdir) which is much better.

File: tests/unit/test_dataset_converter.py [hunk 9]
```diff
@@ -8,15 +8,35 @@
...
+    nwb_paths = [minimal_nwbfile_path.parent]
+    dataset_converter = nwb2bids.DatasetConverter.from_nwb_paths(nwb_paths=nwb_paths)
+
     assert isinstance(dataset_converter, nwb2bids.DatasetConverter)
+    assert isinstance(dataset_converter.session_converters, list)
+    assert len(dataset_converter.session_converters) == 1
+    assert isinstance(dataset_converter.session_converters[0], nwb2bids.SessionConverter)
+
+
+def test_dataset_converter_file_paths_initialization(
```
Comment (candleindark) [inline tests/unit/test_dataset_converter.py:23]: What about adding a test for testing `nwb_paths` containing both file and direcotry paths?

File: tests/unit/test_directory_conditions.py [hunk 10]
```diff
@@ -36,23 +36,23 @@ def test_allowed_directory_conditions(
...
+        dataset_converter = nwb2bids.DatasetConverter.from_nwb_paths(
+            nwb_paths=nwb_paths, additional_metadata_file_path=additional_metadata_file_path
         )
         dataset_converter.extract_metadata()
 
         dataset_converter.write_dataset_description(bids_directory=bids_directory)
     finally:
-        if "implicit" in test_case:
-            os.chdir(initial_working_directory)
+        os.chdir(initial_working_directory)
```
Comment (candleindark) [inline tests/unit/test_directory_conditions.py:55]: See above comment about using monkeypatch.

File: tests/unit/test_directory_conditions.py [hunk 11]
```diff
@@ -110,12 +110,12 @@ def test_disallowed_directory_conditions(
...
+            dataset_converter = nwb2bids.DatasetConverter.from_nwb_paths(
+                nwb_paths=nwb_paths, additional_metadata_file_path=additional_metadata_file_path
             )
             dataset_converter.extract_metadata()
 
             dataset_converter.write_dataset_description(bids_directory=bids_directory)
     finally:
-        if "implicit" in test_case:
-            os.chdir(initial_working_directory)
+        os.chdir(initial_working_directory)
```
Comment (candleindark) [inline tests/unit/test_directory_conditions.py:121]: See the other comment about using monkeypatch

File: tests/unit/test_session_converter.py [hunk 12]
```diff
@@ -8,8 +8,22 @@
...
+    nwb_paths = [minimal_nwbfile_path.parent]
+    session_converters = nwb2bids.SessionConverter.from_nwb_paths(nwb_paths=nwb_paths)
+
+    assert isinstance(session_converters, list)
+    assert len(session_converters) == 1
+    assert isinstance(session_converters[0], nwb2bids.SessionConverter)
+
+
+def test_session_converter_file_path_initialization(
+    minimal_nwbfile_path: pathlib.Path, temporary_bids_directory: pathlib.Path
```
Comment (candleindark) [inline tests/unit/test_session_converter.py:23]: Add a test to test `nwb_paths` being a mix of file and directory paths?

**Review by candleindark — COMMENTED** [review]
Added a new suggestions, nothing major. There is a comment about an edge case which may deserve a separate issue/PR.

File: src/nwb2bids/_command_line_interface/_main.py [hunk 13]
```diff
@@ -46,18 +47,23 @@ def _nwb2bids_cli():
...
+    nwb_paths: tuple[str, ...],
     bids_directory: str | None = None,
     file_mode: typing.Literal["copy", "move", "symlink", "auto"] = "auto",
     additional_metadata_file_path: str | None = None,
 ) -> None:
     """
     Convert NWB files to BIDS format.
 
-    NWB_DIRECTORY : The path to the folder containing NWB files.
+    NWB : Either a path to a folder containing NWB files, or an explicit list of NWB file paths.
```
[the remark this answers is not in the document]
Comment (CodyCBakerPhD) [reply src/nwb2bids/_command_line_interface/_main.py:58]: Feel free to make PR suggestions for easy acceptance

File: src/nwb2bids/_converters/_session_converter.py [hunk 14]
```diff
@@ -33,39 +28,48 @@ class SessionConverter(BaseConverter):
...
-        nwb_file_path_to_session_id = {
-            nwb_file_path: pynwb.read_nwb(nwb_file_path).session_id for nwb_file_path in nwb_file_paths
+        all_nwbfile_paths = []
+        for nwb_path in nwb_paths:
+            if nwb_path.is_file():
+                all_nwbfile_paths.append(nwb_path)
+            elif nwb_path.is_dir():
+                all_nwbfile_paths += list(nwb_path.rglob(pattern="*.nwb"))
+
+        nwbfile_path_to_session_id = {
```
[the remark this answers is not in the document]
Comment (CodyCBakerPhD) [reply src/nwb2bids/_converters/_session_converter.py:56]: Sure, see https://github.com/con/nwb2bids/issues/58

File: tests/unit/test_directory_conditions.py (hunk 11 above)
Comment (CodyCBakerPhD) [reply tests/unit/test_directory_conditions.py]: Swapped

File: tests/unit/test_directory_conditions.py (hunk 10 above)
Comment (CodyCBakerPhD) [reply tests/unit/test_directory_conditions.py]: Swapped

File: tests/integration/test_convert_nwb_dataset.py (hunk 8 above)
Comment (CodyCBakerPhD) [reply tests/integration/test_convert_nwb_dataset.py]: Sure, swapped to monkeypatch

File: tests/unit/test_session_converter.py (hunk 12 above)
Comment (CodyCBakerPhD) [reply tests/unit/test_session_converter.py:23]: Done

**Comment by CodyCBakerPhD** [comment]
@candleindark  Thank you - all comments are now addressed
